### PR TITLE
Use intra-doc links in `str` and `BTreeSet`

### LINKED ISSUE #32129: Broken links in `String` documentation from `Deref` impl
There are a bunch of links in the [`String` documentation](http://doc.rust-lang.org/nightly/std/string/struct.String.html) page which are broken. The reason for this is that most links are drawn from the documentation of the `str` type at `/std/primitive.str.html` but the `String` page is located at `/std/string/struct.String.html`. This mismatch in depth of paths causes a number of broken links, for example on the [`String::split`](http://doc.rust-lang.org/nightly/std/string/struct.String.html#method.split) method.

Not sure of the best way to fix this, but needs a tracking issue.

### LINKED ISSUE #32130: Broken links in BTreeSet struct documentation
There are broken links when looking at the `BTreeSet` structure. This happens because the structure can be viewed from two locations, both at varying depths of links so the relative links may or may not work:
- [`std::collections::BTreeSet`](http://doc.rust-lang.org/nightly/std/collections/struct.BTreeSet.html) - working
- [`std::collections::btree_set::BTreeSet`](http://doc.rust-lang.org/nightly/std/collections/btree_set/struct.BTreeSet.html) - not working

### PR DESCRIPTION
**Pull request by Manishearth**
Fixes #32129, fixes  #32130

A _slight_ degradation in quality is that the `#method.foo` links would previously link to the same page on `String`'s documentation, and now they will navigate to `str`. Not a big deal IMO, and we can also try to improve that.

### CODE REVIEW

File: src/liballoc/collections/btree/set.rs [hunk 1]
```diff
@@ -97,8 +95,7 @@ impl<T: fmt::Debug> fmt::Debug for Iter<'_, T> {
 /// This `struct` is created by the [`into_iter`] method on [`BTreeSet`]
 /// (provided by the `IntoIterator` trait). See its documentation for more.
 ///
-/// [`BTreeSet`]: struct.BTreeSet.html
-/// [`into_iter`]: struct.BTreeSet.html#method.into_iter
+/// [`into_iter`]: BTreeSet#method.into_iter
```
Comment (jyn514) [inline src/liballoc/collections/btree/set.rs:98]: I think this should work:
```suggestion
/// [`into_iter`]: BTreeSet::into_iter
```

File: src/libcore/str/mod.rs [hunk 2]
```diff
@@ -143,7 +142,7 @@ impl FromStr for bool {
 
 /// An error returned when parsing a `bool` using [`from_str`] fails
 ///
-/// [`from_str`]: ../../std/primitive.bool.html#method.from_str
+/// [`from_str`]: FromStr::from_str
```
Comment (jyn514) [inline src/libcore/str/mod.rs:145]: Do we support `<bool as FromStr>::from_str`? If not that seems like a good addition, we could also overload `<bool as FromStr>` to link to the trait implementation. No need to do anything for this PR though.

File: src/libcore/str/mod.rs [hunk 3]
```diff
@@ -266,8 +265,7 @@ impl Utf8Error {
 /// that it is valid UTF-8. `from_utf8()` checks to ensure that the bytes are valid
 /// UTF-8, and then does the conversion.
 ///
-/// [`&str`]: ../../std/primitive.str.html
-/// [`u8`]: ../../std/primitive.u8.html
+/// [`&str`]: str
```
Comment (jyn514) [inline src/libcore/str/mod.rs:268]: Maybe we should strip references automatically? Not sure how that would play with slices eventually though.

File: src/libcore/str/mod.rs [hunk 4]
```diff
@@ -2347,9 +2336,7 @@ impl str {
     }
 
     /// Converts a string slice to a byte slice. To convert the byte slice back
-    /// into a string slice, use the [`str::from_utf8`] function.
-    ///
-    /// [`str::from_utf8`]: ./str/fn.from_utf8.html
+    /// into a string slice, use the [`from_utf8`] function.
```
Comment (jyn514) [inline src/libcore/str/mod.rs:2339]: Let's keep the same link text.
```suggestion
    /// into a string slice, use the [`str::from_utf8`](from_utf8) function.
```

File: src/libcore/str/mod.rs [hunk 5]
```diff
@@ -2429,8 +2416,7 @@ impl str {
     /// The caller must ensure that the returned pointer is never written to.
     /// If you need to mutate the contents of the string slice, use [`as_mut_ptr`].
     ///
-    /// [`u8`]: primitive.u8.html
-    /// [`as_mut_ptr`]: #method.as_mut_ptr
+    /// [`as_mut_ptr`]: str::method.as_mut_ptr
```
Comment (jyn514) [inline src/libcore/str/mod.rs:2419]:
```suggestion
    /// [`as_mut_ptr`]: str::as_mut_ptr
```

File: src/libcore/str/mod.rs [hunk 6]
```diff
@@ -2954,7 +2932,7 @@ impl str {
     ///
     /// To split by Unicode `Whitespace` instead, use [`split_whitespace`].
     ///
-    /// [`split_whitespace`]: #method.split_whitespace
+    /// [`split_whitespace`]: str::method.split_whitespace
```
Comment (jyn514) [inline src/libcore/str/mod.rs:2935]:
```suggestion
    /// [`split_whitespace`]: str::split_whitespace
```

File: src/libcore/str/mod.rs [hunk 7]
```diff
@@ -3068,8 +3046,7 @@ impl str {
     /// The [pattern] can be a `&str`, [`char`], a slice of [`char`]s, or a
     /// function or closure that determines if a character matches.
     ///
-    /// [`char`]: primitive.char.html
-    /// [pattern]: str/pattern/index.html
+    /// [pattern]: self::pattern
```
Comment (jyn514) [inline src/libcore/str/mod.rs:3049]: Hmm, this is not very clear that it's the module and not an associated function ... but I guess as long as the link works it's fine.

**Review by jyn514 — COMMENTED** [review]
This reminds me why I love intra-doc links! This looks great :) There are a lot of broken links though. I guess the link checker won't check those until https://github.com/rust-lang/rust/pull/71670/ lands, can we wait for that?

**Comment by jyn514** [comment]
Hmm, well it caught at least some of them.
```
 std/primitive.str.html:128: broken link - std/str::method.as_mut_ptr
std/primitive.str.html:434: broken link - std/str::method.split_whitespace
std/primitive.str.html:1212: broken link - std/str::method.to_ascii_lowercase
thread 'main' panicked at 'found some broken links', src/tools/linkchecker/main.rs:41:9
```
but I would still prefer to wait for https://github.com/rust-lang/rust/pull/71670.

File: src/liballoc/collections/btree/set.rs (hunk 1 above)
Comment (Manishearth) [reply src/liballoc/collections/btree/set.rs:98]: Nah, it's a trait method

File: src/libcore/str/mod.rs (hunk 2 above)
Comment (Manishearth) [reply src/libcore/str/mod.rs:145]: We do not

File: src/libcore/str/mod.rs (hunk 7 above)
Comment (Manishearth) [reply src/libcore/str/mod.rs:3049]: It is, associated functions are `Self`

**Comment by Manishearth** [comment]
Fixed the links. The dot made us treat it as a URL and skip the lint (which is expected).

**Comment by jyn514** [comment]
Oh huh, I forgot that @lzutao turned on `deny(resolution_failure)` (thank you ❤️). So we should be good merge once you fix the broken links.

```
 error: `[str::to_ascii_lowercase]` cannot be resolved, ignoring it.
    --> src/libcore/str/mod.rs:4346:33
     |
4346 |     /// [`to_ascii_lowercase`]: str::to_ascii_lowercase
     |                                 ^^^^^^^^^^^^^^^^^^^^^^^ cannot be resolved, ignoring
     |
note: the lint level is defined here
    --> src/libcore/lib.rs:64:9
     |
64   | #![deny(intra_doc_link_resolution_failure)] // rustdoc is run without -D warnings
     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     = help: to escape `[` and `]` characters, just add '\' before them like `\[` or `\]`

error: aborting due to previous error
```

**Comment by Manishearth** [comment]
Fixed. We can't link to any libstd things from libcore, so I reverted those.

**Comment by jyn514** [comment]
r=me once CI is passing

**Comment by Manishearth** [comment]
@bors r=jyn514

that's what bors is for :smile: (and when making rollups we typically make sure CI is passing)

**Comment by ehuss** [comment]
This PR introduces quite a few regressions (as also noted in #74061) which I think would be unfortunate to add. In particular, many of the links are switched to remote links to https://doc.rust-lang.org/nightly/.  Also, though not as severe, many of the primitive links now go to the module of the same name.

**Comment by Manishearth** [comment]
@ehuss I don't consider it a regression for the links to link to doc.rlo. The primitive links can be fixed if we want.

**Comment by Manishearth** [comment]
FWIW in my local build the links do not link to doc.rlo, they link locally. I think what you're seeing is https://github.com/rust-lang/rust/issues/74458

**Comment by ehuss** [comment]
> I don't consider it a regression for the links to link to doc.rlo.

Why not? A fair bit of effort goes into making sure the documentation is usable offline. I also find it confusing when browsing the stable docs, then ending up (without warning) in the nightly docs, using something that appears to be stable, only to learn much later that it is not stable (or doesn't exist at all on stable).

To be clear, it is better to fix the broken links on the String page, but I consider it a backstep in other regards that should (eventually) be fixed.

> FWIW in my local build the links do not link to doc.rlo, they link locally. I think what you're seeing is #74458

An example is `core/str/index.html`. Almost all of the links in the descriptions there now go to the remote site. I'm guessing that core and alloc are unable to link to the primitive types?  Perhaps that will need to be special cased?

**Comment by Manishearth** [comment]
@ehuss That's only on the libcore docs, it's because libcore is weird, yes. We might be able to fix that, I'll file an issue.

**Comment by Manishearth** [comment]
Shit, closed the PR by mistake

**Comment by Manishearth** [comment]
Opened as https://github.com/rust-lang/rust/pull/74453

**Comment by Manishearth** [comment]
@ehuss oh, right, this is because primitives only exist in `std`, and when documenting core we don't have `std` yet.

**Comment by jyn514** [comment]
> primitives only exist in std

That seems weird, shouldn't they be in libcore?

**Pull request closed without merging**